### ISSUE TITLE
Suggestion: approach network games with dedicated server

### ISSUE THREAD

**Issue by Leonardo-Ferreira** [OpenSAGE/OpenSAGE#508]
Hi all,
I can see that so far you have not implemented a "dedicated server" code, but you have made some progress on LAN playing. I would like to suggest another "interpretation" on the problem: all skirmish/lan games are actually a dedicated server backed game.
The idea is simple: whenever one of these start (skirmish or lan game) you would also start another process which would be the dedicated server.

The local dedicated network server would be more resilient because it would be independent from the client (would have its own memory set and GC for example). Also, a crash from the client wouldn't mean "Game Over" because the server process would still be running, allowing for the "host" to rejoin (in the future I know)...
A skirmish game would not be harmed because the loopback address is actually reliable...

well, that's my suggestion, good luck on your endeavour!

**Comment by EndlessEden** [comment]
I agree with this idea, however, there should be some type of save on unsafe socket closure method added to the server. so in the event of a client crash, with a game started on localhost, it automatically saves the running session state to file, and gracefully kills the server session.

implementing a save/reload state should be rather simple i would think.

**Comment by feliwir** [comment]
We kind of discussed this issue already on #34 . The current idea is to stay with peer to peer lockstep, but use a dedicated server for the networking lobby. 

That way we don't have to care about hosting a large infrastructure (which we can't do) and the overall amount of complexity is probably lower, since LAN/Online/Replays all share the same code.  But please take a look at the linked issue above to see all the details for our current decision.

**Comment by Leonardo-Ferreira** [comment]
@feliwir yeah, that's a long thread, that got lost in anti-cheating...

Anyhow, I would just like to point that I believe that you're working on false-premises when dealing with the dedicated server issue. 
First false premisse, that **YOU** have to host something. Whoever's interested on multiplayer gaming can set up his own dedicated server, and people will do if it's easy enough. Nowadays can be as easy as creating a cloud account and creating a VM with a pre-defined image provided by the project...
Second misconception is that it **HAS** to be large. That depends solely on how performant/economic the code is. You'd be amazed with the kind of things that are backed by single core-2gb ram with SSD vm. (currently $28/month on AWS EC2 Virginia... and yes, billed by the minute, so turn it off, stop paying)
Third is that the `overall amount of complexity is probably lower`. When you have the "server" and the client running in the same AppDomain, debugging is difficulty is higher. You'd also have a higher difficulty keep the separation of concerns. Troubleshooting P2P communication IS A-F*****-LOT more difficult. Desync issues are HARD to track/trace/report and to solve.

In the end, I don't think the P2P decision was actually taken (you guys kinda rolled with it) and it inherently obligates the lockstep model because the bandwidth required could be a problem for large games, not to mention games with spectators (which I believe is in the roadmap).
During combat a player can realistically move-around/orchestrate up to 100 units depending on how they are grouped... imagine a combat between 2-3 players on 6v6 game where the others are either moving into the battle zone or flanking other opponents... Even with lockstep this will be though, now throw some latency in there and you have 12 frustrated players with a desynced/crashed game...

**Comment by ibraheemhlaiyil** [comment]
+1 on a server, I think OpenSAGE deserves a modern take on this. The server can be spun up on any player's machine too if they can forward a port on their router to their local machine. How about 2-way websockets?

**Comment by Leonardo-Ferreira** [comment]
>How about 2-way websockets

another valid approach! gRPC can also fit the bill.
